samza-core/src/main/scala/org/apache/samza/system/SystemConsumers.scala

```diff
@@ -192,10 +194,22 @@ class SystemConsumers (
 
     metrics.registerSystemStreamPartition(systemStreamPartition)
     unprocessedMessagesBySSP.put(systemStreamPartition, new ArrayDeque[IncomingMessageEnvelope]())
+
+    // Note regarding Startpoints and MessageChooser:
+    // Even if there is a startpoint for this SSP, passing in the checkpoint offset should not have any side-effects.
+    // Basically, the offset in the chooser is used in the special scenario where an SSP is both a broadcast and bootstrap stream
+    // and needs to decide what's the lowest starting offset for an SSP that spans across multiple tasks so it knows
+    // to keep the highest priority on the SSP starting from the lowest starting offset until the SSP is fully
+    // bootstrapped to the UPCOMING offset. The offset here is ignored otherwise.
     chooser.register(systemStreamPartition, offset)
 
     try {
-      consumers(systemStreamPartition.getSystem).register(systemStreamPartition, offset)
+      val consumer = consumers(systemStreamPartition.getSystem)
+      if (startpoint != null && consumer.isInstanceOf[StartpointVisitor]) {
+        startpoint.apply(systemStreamPartition, consumer.asInstanceOf[StartpointVisitor])
+      } else {
+        consumer.register(systemStreamPartition, offset)
```

Comment:
Could this put the `offset` into a `StartpointSpecific` and then just always call `consumer.register(systemStreamPartition, startpoint)`?